roofit/roofitcore/src/RooGradMinimizerFcn.cxx

```diff
@@ -0,0 +1,294 @@
+/*****************************************************************************
+ * Project: RooFit                                                           *
+ * Package: RooFitCore                                                       *
+ * @(#)root/roofitcore:$Id$
+ * Authors:                                                                  *
+ *   AL, Alfio Lazzaro,   INFN Milan,        alfio.lazzaro@mi.infn.it        *
+ *   PB, Patrick Bos,     NL eScience Center, p.bos@esciencecenter.nl        *
+ *   VC, Vince Croft,     DIANA / NYU,        vincent.croft@cern.ch          *
+ *                                                                           *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// RooGradMinimizerFcn is am interface class to the ROOT::Math function
+// for minimization. See RooGradMinimizer.cxx for more information.
+//
+
+#include "RooGradMinimizerFcn.h"
+
+#include <iostream>
+
+#include "Riostream.h"
+
+#include "TIterator.h"
+#include "TClass.h"
+
+#include "RooAbsArg.h"
+#include "RooAbsPdf.h"
+#include "RooArgSet.h"
+#include "RooRealVar.h"
+#include "RooMsgService.h"
+
+#include "RooMinimizer.h"
+
+#include "Fit/Fitter.h"
+#include "Math/Minimizer.h"
+
+#include <algorithm> // std::equal
+
+RooGradMinimizerFcn::RooGradMinimizerFcn(RooAbsReal *funct, RooMinimizer *context, bool verbose)
+   : RooAbsMinimizerFcn(RooArgList(*funct->getParameters(RooArgSet())), context, verbose),
```

Comment:
```suggestion
   : RooAbsMinimizerFcn(RooArgList(getParametersForRooAbsMinimizerFcn(*funct), context, verbose),
```
There is a memory leak here, because `getParameters` returns an owning pointer.
I would write a little helper function and put it into the anonymous namespace a few lines above:
```C++
namespace {
   RooArgList getParametersForRooAbsMinimizerFcn(RooAbsArg const& arg) {
      RooArgSet out; 
      arg.getParameters(nullptr, out);    
      return RooArgList(out);
   }
} // namespace
```

Reply:
Bit weird alternative, but gets the job done in one line:
```suggestion
   : RooAbsMinimizerFcn(RooArgList( * std::unique_ptr<RooArgSet>(funct->getParameters({})) ), context, verbose),
```
Maybe the `{}` doesn't work to construct a RooArgSet, in this case, you have to revert that part.

Reply:
Would have been nice if `bool RooAbsArg::getParameters` would just return a `RooArgSet`. The `bool` seems to always return `false` anyway. Theoretically in some override, but the only override I can find is in `RooBarlowBeestonLL.cxx`, and there it returns `false || false` 😄

Anyway, that's outside of the scope of this PR.

I'm gonna go with the weird one-liner then, if you are both ok with that :)